Add Intro_DP Focus Problem (#1885)
cpinitiative/usaco-guide: content/4_Gold/Intro_DP.mdx

---
id: intro-dp
title: 'Introduction to DP'
author: Michael Cao, Benjamin Qi, Neo Wang
prerequisites:
  - complete-rec
  - modular
description: 'Speeding up naive recursive solutions with memoization.'
frequency: 4
---

> 

Reviewer: ```suggestion
```

> <FocusProblem problem='focus' />

Reviewer: ```suggestion
```

with dynamic programming by keeping track of "optimal states" in order to avoid
such expensive calculations - in this case, the minimum cost to reach stone $i$.

### With Dynamic Programming

**Time Complexity:** $\mathcal{O}(N)$

There are only two options: jumping once, or jumping twice. Define
$\texttt{dp}[i]$ as the minimum cost to reach stone $i$. Therefore,
$\texttt{dp}[i+1]$ must represent the next stone, and $\texttt{dp}[i+2]$ must
represent the stone after that. Then, our transitions are as follows at stone
$i$ must be:

- Jump one stone, incurring a cost of $|h_i - h_{i+1}|$:

  $$
  \texttt{dp}[i + 1] = \min(\texttt{dp}[i + 1], \texttt{dp}[i] + |h_i - h_{i + 1}|)
  $$

- Jump two stones, incurring a cost of $|h_i - h_{i + 2}|$:
  $$
  \texttt{dp}[i + 2] = \min(\texttt{dp}[i + 2], \texttt{dp}[i] + |h_i - h_{i + 2}|)
  $$

We can start with the base case that $\texttt{dp}[1] = 0$, since the frog is
already on that square, and proceed to calculate
$\texttt{dp}[1], \texttt{dp}[2], \ldots \texttt{dp}[N]$. Note that in the code
we ignore $\texttt{dp}[i]$ if $i>N$.

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

using namespace std;

> const int MAX_N = 1e5+1;

Reviewer 2: ```suggestion
const int MAX_N = 1e5 + 1;
```

> int H[MAX_N], dp[MAX_N];

Reviewer 2: no 1 letter variable names D:

</Info>

## Solution - Frog 1

The problem asks us to compute the minimum total cost it takes for a frog to
travel from stone $1$ to stone $N (N \le 10^5)$ given that the frog can only
jump a distance of one or two. The cost to travel between any two stones $i$ and
$j$ is given by $|h_i - h_j|$, where $h_i$ represents the height of stone $i$.

### Without Dynamic Programming

**Time Complexity:** $\mathcal{O}(2^N)$

Since there are only two options, we can use recursion to compute what would
happen if we jumped either $1$ stone, or $2$ stones. There are two
possibilities, so recursively computing would require computing both a left and
right subtree. Therefore, for every additional jump, each branch splits into
two, which results in an exponential time complexity. However, this can sped up
with dynamic programming by keeping track of "optimal states" in order to avoid

> such expensive calculations - in this case, the minimum cost to reach stone $i$.

Reviewer 3: add that you do this because you don't want to recalculate the same state

### With Dynamic Programming

**Time Complexity:** $\mathcal{O}(N)$

There are only two options: jumping once, or jumping twice. Define
$\texttt{dp}[i]$ as the minimum cost to reach stone $i$. Therefore,
$\texttt{dp}[i+1]$ must represent the next stone, and $\texttt{dp}[i+2]$ must
represent the stone after that. Then, our transitions are as follows at stone
$i$ must be:

> - Jump one stone, incurring a cost of $|h_i - h_{i+1}|$:

Reviewer 2: could use height[i] to match the code

</Info>

## Solution - Frog 1

The problem asks us to compute the minimum total cost it takes for a frog to
travel from stone $1$ to stone $N (N \le 10^5)$ given that the frog can only
jump a distance of one or two. The cost to travel between any two stones $i$ and
$j$ is given by $|h_i - h_j|$, where $h_i$ represents the height of stone $i$.

### Without Dynamic Programming

**Time Complexity:** $\mathcal{O}(2^N)$

Since there are only two options, we can use recursion to compute what would
happen if we jumped either $1$ stone, or $2$ stones. There are two
possibilities, so recursively computing would require computing both a left and
right subtree. Therefore, for every additional jump, each branch splits into

> two, which results in an exponential time complexity. However, this can sped up

Reviewer 4: ```suggestion
two, which results in an exponential time complexity.

However, this can be sped up
```

with dynamic programming by keeping track of "optimal states" in order to avoid

> redundant calculations. For example, recursively calculating jumps of length

Reviewer 4: ```suggestion
calculating states multiple times. For example, recursively calculating jumps of length
```

</Info>

> ## Solution - Frog 1

Reviewer: ```suggestion
## Example - Frog 1

<FocusProblem problem='focus' />
```

</Info>

## Solution - Frog 1

> **Time Complexity:** $\mathcal{O}(N)$

Reviewer 3: describe a solution w/o dp and how dp speeds it up

to avoid such expensive calculations - in this case, the minimum cost to
reach stone $i$.

### With Dynamic Programming

**Time Complexity:** $\mathcal{O}(N)$

There are only two options: jumping once, or jumping twice. Define
$\texttt{dp}[i]$ as the minimum cost to reach stone $i$. Therefore,
$\texttt{dp}[i+1]$ must represent the next stone, and $\texttt{dp}[i+2]$ must
represent the stone after that. Then, our transitions are as follows at stone
$i$ must be:

- Jump one stone, incurring a cost of $|h_i - h_{i+1}|$:

  $$
  \texttt{dp}[i + 1] = \min(\texttt{dp}[i + 1], \texttt{dp}[i] + |\texttt{H}[i] - \texttt{H}[i + 1]|)
  $$

- Jump two stones, incurring a cost of $|h_i - h_{i + 2}|$:
  $$
  \texttt{dp}[i + 2] = \min(\texttt{dp}[i + 2], \texttt{dp}[i] + |\texttt{H}[i] - \texttt{H}[i + 2]|)
  $$

We can start with the base case that $\texttt{dp}[1] = 0$, since the frog is
already on that square, and proceed to calculate
$\texttt{dp}[1], \texttt{dp}[2], \ldots \texttt{dp}[N]$. Note that in the code
we ignore $\texttt{dp}[i]$ if $i>N$.

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

using namespace std;

const int MAX_N = 1e5+1;

> int H[MAX_N], dp[MAX_N];

Reviewer 5: what's H

Author: the height of the stone

int main() {
	cin.tie(0)->sync_with_stdio(0);

	int N; cin >> N;

> 

Reviewer 5: ```suggestion
```

	for(int i = 1; i <= N; ++i) {
		cin >> H[i];
		dp[i] = 1e9 + 7;
	}

	dp[1] = 0;

> 

Reviewer 5: ```suggestion
```

	for(int i = 1; i <= N; ++i) {
		if(i + 1 <= N) dp[i + 1] = min(dp[i + 1], dp[i] + abs(H[i] - H[i + 1]));
		if(i + 2 <= N) dp[i + 2] = min(dp[i + 2], dp[i] + abs(H[i] - H[i + 2]));
	}

> 

Reviewer 5: ```suggestion
```
